curvefs/src/mds/space/usage.h

```diff
@@ -0,0 +1,80 @@
+/*
+ *  Copyright (c) 2022 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Date: Friday Mar 11 16:50:20 CST 2022
+ * Author: wuhanqing
+ */
+
+#ifndef CURVEFS_SRC_MDS_SPACE_USAGE_H_
+#define CURVEFS_SRC_MDS_SPACE_USAGE_H_
+
+#include <atomic>
+#include <cstdint>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "curvefs/proto/space.pb.h"
+#include "src/common/concurrent/rw_lock.h"
+
+namespace curvefs {
+namespace mds {
+namespace space {
+
+using ::curve::common::BthreadRWLock;
+
+class SpaceUsage {
+ public:
+    SpaceUsage(uint64_t size,
+               uint64_t availableSize,
+               uint64_t blockGroupSize,
+               const std::unordered_map<uint64_t, BlockGroup>& allocatedGroups);
+
+    void OnAllocateGroups(const std::vector<BlockGroup>& groups);
+
+    void OnReleaseGroups(const std::vector<BlockGroup>& groups);
+
+    void OnUpdate(const std::vector<BlockGroup>& groups);
+
+    void OnExtentVolume(uint64_t extentBytes, uint64_t extentBlockGroups);
+
+    void Stat(SpaceInfo* info);
+
+ private:
+    struct BlockGroupUsage {
+        uint64_t available;
+        std::string owner;
+    };
+
+    const uint64_t blockGroupSize_;
+
+    std::atomic<uint64_t> totalSize_;
```

Comment:
Defined as a metric for easy access to statistical information

Reply:
ditto